### PR TITLE
Refactor runtime selection and checkFirebaseSDKVersion

### PR DESCRIPTION
**Pull request by laurenzlong**
### Description

- Fix `checkFirebaseSDKVersion`, which no longer works reliably since `npm outdated` often returns empty result in newer versions of NPM (https://github.com/npm/npm/issues/16840). Instead, I refactored it to use `npm list` and `npm show` and compare the 2 results.
- Removed a bunch of redundant code for getting runtime choice, since it was being done in multiple places, and also we now require that package.json includes "engines" field so there's no more need to have defaults, etc. (Context: https://github.com/firebase/firebase-tools/pull/1220)
- Renamed `runtimeChoiceSelector.js` to `parseRuntimeAndValidateSDK.ts` to be more descriptive. It took me a long time to figure out that was the file that set "minimum SDK versions". Its `functionsSDKTooOld` also needed to be fixed since it only took into account `package.json` and not `package-lock.json` (It now calls the `getFunctionsSDKVersion` method in `checkFirebaseSDKVersion.ts`, which makes use of `npm list`, and that takes `package-lock.json` into account.

### Scenarios Tested
Node 6 warning
node6

<2.0 SDK warning
Screen Shot 2020-04-14 at 11 18 11 AM

### CODE REVIEW

File: src/checkFirebaseSDKVersion.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
+"use strict";
+
+import * as _ from "lodash";
+import * as clc from "cli-color";
+import * as path from "path";
+import * as semver from "semver";
+import { spawnSync } from "child_process";
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts]: I saw this change and figured I'd bring it up now (don't know if this is ready for review yet): `cross-spawn` is used so that processes can be spun up in Windows and Linux environments. The built-in handlers were having trouble with this.

Comment (laurenzlong) [reply src/checkFirebaseSDKVersion.ts]: This PR is still in draft, but good to know thanks!

Comment (laurenzlong) [reply src/checkFirebaseSDKVersion.ts]: Changed back to spawnSync

File: src/checkFirebaseSDKVersion.ts [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    }
+    const output: NpmListResult = JSON.parse(child.stdout);
+    return _.get(output, ["dependencies", "firebase-functions", "version"]);
+  } catch (e) {
+    logger.debug("getFunctionsSDKVersion encountered error:", e);
+    return;
+  }
+}
+
+export async function checkSDKVersion(options: any): Promise<void> {
```
Comment (joehan) [inline src/checkFirebaseSDKVersion.ts:46]: nit - rename to checkFunctionsSDKVersion for consistency

File: src/parseRuntimeAndValidateSDK.ts [hunk 3]
```diff
@@ -22,17 +24,20 @@ const ENGINE_RUNTIMES: { [key: string]: string } = {
...
   `package.json in functions directory has an engines field which is unsupported. ` +
     `The only valid choices are: ${clc.bold('{"node": "8"}')} and ${clc.bold('{"node": "10"}')}. ` +
     `Note that Node.js 6 is now deprecated.`
 );
-export const DEPRECATION_WARNING_MSG =
+
+export const NODE6_DEPRECATION_WARNING_MSG =
   clc.bold.yellow("functions: ") +
-  "Deploying functions to Node 6 runtime, which is deprecated. Node 8 is available " +
+  "Deploying functions to Node 6 runtime, which is deprecated. Node 10 is available " +
```
Comment (joehan) [inline src/parseRuntimeAndValidateSDK.ts:35]: Optional - https://cloud.google.com/functions/docs/concepts/nodejs-6-runtime says 4-22-2020 is when GCF will fully stop supporting Node 6. It might be more appropriate top do all the cleanup in one separate PR, but it would be nice to add that date to this warning message.

File: src/test/parseRuntimeAndValidateSDK.spec.ts [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.eventually.deep.equal("nodejs8");
+    expect(warningSpy).calledWith(runtime.NODE8_DEPRECATION_WARNING_MSG);
+  });
+
+  it("should return node 10 if package.json engines field is set to node 10", () => {
+    cjsonStub.returns({ engines: { node: "10" } });
+    SDKVersionStub.returns("3.4.0");
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.eventually.deep.equal("nodejs10");
```
Comment (joehan) [inline src/test/parseRuntimeAndValidateSDK.spec.ts:53]: Is deep equal necessary for comparing strings here?

File: src/test/parseRuntimeAndValidateSDK.spec.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.eventually.deep.equal("nodejs6");
+    expect(warningSpy).calledWith(runtime.NODE6_DEPRECATION_WARNING_MSG);
+  });
+
+  it("should return node 8 if package.json engines field is set to node 8 and print warning", () => {
+    cjsonStub.returns({ engines: { node: "8" } });
+    SDKVersionStub.returns("2.0.0");
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.eventually.deep.equal("nodejs8");
```
Comment (joehan) [inline src/test/parseRuntimeAndValidateSDK.spec.ts:45]: Is deep equal necessary for comparing strings here?

File: src/test/parseRuntimeAndValidateSDK.spec.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  it("should return node 6 if package.json engines field is set to node 6 and print warning", () => {
+    cjsonStub.returns({ engines: { node: "6" } });
+    SDKVersionStub.returns("2.0.0");
+
+    expect(runtime.getRuntimeChoice("path/to/source")).to.eventually.deep.equal("nodejs6");
```
Comment (joehan) [inline src/test/parseRuntimeAndValidateSDK.spec.ts:37]: Is deep equal necessary for comparing strings here?

File: src/test/parseRuntimeAndValidateSDK.spec.ts [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    cjsonStub.returns({ engines: { node: "10" } });
+    SDKVersionStub.returns("0.5.0");
+
+    await runtime.getRuntimeChoice("path/to/source");
+    expect(warningSpy).calledWith(runtime.FUNCTIONS_SDK_VERSION_TOO_OLD_WARNING);
+  });
+
+  it("should not throw error if user's SDK version fails to be fetched", () => {
+    cjsonStub.returns({ engines: { node: "10" } });
+    // Intentionally not setting SDKVersionStub.
```
Comment (joehan) [inline src/test/parseRuntimeAndValidateSDK.spec.ts:67]: Nit - why are you intentionally not setting this?

**Comment by karayu** [comment]
LGTM. Lets keep the DO NOT Merge for now.

**Comment by bkendall** [comment]
Happy to take a look at this again when it's ready. Let me know!

**Comment by laurenzlong** [comment]
@bkendall It's ready! Thanks.

File: src/checkFirebaseSDKVersion.ts (hunk 2 above)
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:46]: With the system calls being `spawn.sync`, I don't think this has to be `async` either

File: src/parseRuntimeAndValidateSDK.ts [hunk 8]
```diff
@@ -57,40 +62,50 @@ export function getHumanFriendlyRuntimeName(runtime: string): string {
  * package.json.
  * @param sourceDir directory where the functions are defined.
  */
-export function getRuntimeChoice(sourceDir: string): any {
+export async function getRuntimeChoice(sourceDir: string): Promise<string> {
```
Comment (bkendall) [inline src/parseRuntimeAndValidateSDK.ts:65]: Following the 'async' functions, I don't think this needs to be async either.

File: src/test/parseRuntimeAndValidateSDK.spec.ts (hunk 6 above)
Comment (bkendall) [inline src/test/parseRuntimeAndValidateSDK.spec.ts:37]: For tests that are async (`eventually` is the part that's async here), mocha expects a promise to be returned. The test function needs to be `async` and it either has to return a single `expect...eventually`, or do `await expect...eventually`

(the failure modes for this are weird because it will look like unhandled promise rejections if this test fails)

File: src/checkFirebaseSDKVersion.ts (hunk 2 above)
Comment (laurenzlong) [reply src/checkFirebaseSDKVersion.ts:46]: done

File: src/test/parseRuntimeAndValidateSDK.spec.ts (hunk 7 above)
Comment (laurenzlong) [reply src/test/parseRuntimeAndValidateSDK.spec.ts:67]: clarified.

File: src/test/parseRuntimeAndValidateSDK.spec.ts (hunk 6 above)
Comment (laurenzlong) [reply src/test/parseRuntimeAndValidateSDK.spec.ts:37]: this method is no longer async, so this no longer applies.

File: src/parseRuntimeAndValidateSDK.ts (hunk 3 above)
Comment (laurenzlong) [reply src/parseRuntimeAndValidateSDK.ts:35]: I'm removing all Node version messaging changes from this PR, in lieu of Michael's PR.

File: src/checkFirebaseSDKVersion.ts [hunk 9]
```diff
@@ -0,0 +1,96 @@
+import * as _ from "lodash";
```
Comment (joehan) [inline src/checkFirebaseSDKVersion.ts:1]: Should we rename this file to checkFunctionsSDKVersion now?

File: src/commands/deploy.js [hunk 10]
```diff
@@ -1,12 +1,11 @@
 "use strict";
 
 const _ = require("lodash");
-
 const requireInstance = require("../requireInstance");
 const { requirePermissions } = require("../requirePermissions");
 const { checkServiceAccountIam } = require("../deploy/functions/checkIam");
 const checkValidTargetFilters = require("../checkValidTargetFilters");
-const checkFirebaseSDKVersion = require("../checkFirebaseSDKVersion");
+var checkFunctionsSDKVersion = require("../checkFirebaseSDKVersion").checkFunctionsSDKVersion;
```
Comment (joehan) [inline src/commands/deploy.js:8]: Any reason this needs to be a var? If possible, a const would be preffered

File: src/parseRuntimeAndValidateSDK.ts [hunk 11]
```diff
@@ -57,15 +57,19 @@ export function getHumanFriendlyRuntimeName(runtime: string): string {
...
-    return null;
-    // TODO(b/129422952): Change to throw error instead of returning null
-    // when engines field in package.json becomes required:
-    // throw new FirebaseError(ENGINES_FIELD_REQUIRED_MSG, { exit: 1 });
+    // We should really never hit this, since deploy/functions/prepare already checked that package.json has an "engines" field.
+    throw new FirebaseError(
+      `Engines field is required but was not found in package.json.\n` +
+        `To fix this, add the following lines to your package.json: \n
+      "engines": {
+        "node": "10"
```
Comment (joehan) [inline src/parseRuntimeAndValidateSDK.ts:70]: More of a product question, but do we want to tell users to go to Node 10?

File: templates/init/functions/javascript/package.lint.json [hunk 12]
```diff
@@ -10,16 +10,16 @@
...
   },
   "engines": {
-    "node": "8"
+    "node": "10"
   },
   "dependencies": {
-    "firebase-admin": "^8.9.0",
-    "firebase-functions": "^3.3.0"
+    "firebase-admin": "^8.10.0",
+    "firebase-functions": "^3.5.0"
```
Comment (joehan) [inline templates/init/functions/javascript/package.lint.json:17]: Should this be ^3.6.0 https://github.com/firebase/firebase-functions/releases/tag/v3.6.0

File: templates/init/functions/javascript/package.nolint.json [hunk 13]
```diff
@@ -9,14 +9,14 @@
...
   },
   "engines": {
-    "node": "8"
+    "node": "10"
   },
   "dependencies": {
-    "firebase-admin": "^8.9.0",
-    "firebase-functions": "^3.3.0"
+    "firebase-admin": "^8.10.0",
+    "firebase-functions": "^3.5.0"
```
Comment (joehan) [inline templates/init/functions/javascript/package.nolint.json:16]: Should this be https://github.com/firebase/firebase-functions/releases/tag/v3.6.0

File: templates/init/functions/typescript/package.lint.json [hunk 14]
```diff
@@ -10,17 +10,17 @@
...
   "engines": {
-    "node": "8"
+    "node": "10"
   },
   "main": "lib/index.js",
   "dependencies": {
-    "firebase-admin": "^8.9.0",
-    "firebase-functions": "^3.3.0"
+    "firebase-admin": "^8.10.0",
+    "firebase-functions": "^3.5.0"
```
Comment (joehan) [inline templates/init/functions/typescript/package.lint.json:18]: Should this be https://github.com/firebase/firebase-functions/releases/tag/v3.6.0

File: templates/init/functions/typescript/package.nolint.json [hunk 15]
```diff
@@ -9,16 +9,16 @@
...
   "engines": {
-    "node": "8"
+    "node": "10"
   },
   "main": "lib/index.js",
   "dependencies": {
-    "firebase-admin": "^8.9.0",
-    "firebase-functions": "^3.3.0"
+    "firebase-admin": "^8.10.0",
+    "firebase-functions": "^3.5.0"
```
Comment (joehan) [inline templates/init/functions/typescript/package.nolint.json:17]: Should this be https://github.com/firebase/firebase-functions/releases/tag/v3.6.0

File: src/parseRuntimeAndValidateSDK.ts (hunk 11 above)
Comment (laurenzlong) [reply src/parseRuntimeAndValidateSDK.ts:70]: That's in the other PR (Michael's)

File: src/checkFirebaseSDKVersion.ts (hunk 9 above)
Comment (laurenzlong) [reply src/checkFirebaseSDKVersion.ts:1]: I want to leave this file as is, since there may be other checks for other SDKs in the future. And this is a general pre-deployment step for "firebase deploy" (in commands/deploy.js), not exclusive to functions.

File: templates/init/functions/javascript/package.lint.json (hunk 12 above)
Comment (laurenzlong) [reply templates/init/functions/javascript/package.lint.json:17]: good catch!

**Comment by laurenzlong** [comment]
I've tested after my changes, screenshots have been updated.

File: src/parseRuntimeAndValidateSDK.ts (hunk 11 above)
Comment (laurenzlong) [reply src/parseRuntimeAndValidateSDK.ts:70]: @mbleigh for input. I think is, because Node 10 is going to GA quite soon, so it's a good default to suggest. (And those that really care about only using GA runtimes will not have an empty engine field in package.json. )

File: templates/init/functions/javascript/package.lint.json (hunk 12 above)
Comment (joehan) [reply templates/init/functions/javascript/package.lint.json:17]: Still on 3.5.0?

**Comment by mbleigh** [comment]
@laurenzlong we should not recommend Node 10 as the default just yet given upcoming infrastructure changes.

File: templates/init/functions/javascript/package.lint.json (hunk 12 above)
Comment (laurenzlong) [reply templates/init/functions/javascript/package.lint.json]: fixed!

File: src/parseRuntimeAndValidateSDK.ts (hunk 11 above)
Comment (laurenzlong) [reply src/parseRuntimeAndValidateSDK.ts:70]: To capture our offline conversation: the init templates will stay at Node 8, but we can update the error messages now.

File: CHANGELOG.md [hunk 16]
```diff
@@ -0,0 +1 @@
+- Fixes bug where firebase-functions SDK version checking was unreliable and did not consider package-lock.json.
```
Comment (bkendall) [inline CHANGELOG.md:1]:
```suggestion
- Fixes a bug where `firebase-functions` SDK version checking was unreliable and did not consider `package-lock.json`.
```

File: src/checkFirebaseSDKVersion.ts [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+import * as clc from "cli-color";
+import * as path from "path";
+import * as semver from "semver";
+import * as spawn from "cross-spawn";
+import * as utils from "./utils";
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:6]: separate local imports from the module imports, please

File: src/checkFirebaseSDKVersion.ts [hunk 18]
```diff
@@ -0,0 +1,96 @@
...
+  };
+}
+
+interface NpmShowResult {
+  "dist-tags": {
+    latest: string;
+  };
+}
+
+/** Returns the version of firebase-functions SDK specified by package.json and package-lock.json.
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:26]: Start this description on the next line.

File: src/checkFirebaseSDKVersion.ts [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+ * @param sourceDir Source directory of functions code
+ **/
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:28]: Double asterisks here - only need one.

Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:28]: Do include a `@return` description

File: src/checkFirebaseSDKVersion.ts [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+  if (!currentVersion) {
+    logger.debug("getFunctionsSDKVersion was unable to retrieve 'firebase-functions' version");
+    return;
+  }
+  try {
+    const child = spawn.sync("npm", ["show", "firebase-functions", "--json=true"], {
+      encoding: "utf8",
+    });
+    if (child.error) {
+      logger.debug(child.error.stack);
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:67]: A message here will help too (along with printing the error/stack)

File: src/checkFirebaseSDKVersion.ts [hunk 21]
```diff
@@ -0,0 +1,96 @@
...
+      return;
+    }
+    const output: NpmShowResult = JSON.parse(child.stdout);
+    if (!output || _.isEmpty(output)) {
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:71]: `isEmpty` returns true if `undefined` is passed to it - is that sufficient?

File: src/checkFirebaseSDKVersion.ts [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+    const output: NpmListResult = JSON.parse(child.stdout);
+    return _.get(output, ["dependencies", "firebase-functions", "version"]);
+  } catch (e) {
+    logger.debug("getFunctionsSDKVersion encountered error:", e);
+    return;
+  }
+}
+
+/**
+ * Checks if firebase-functions SDK is not the latest version in NPM, and prints update notice if it is outdated.
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:48]: Given the way this is written, I would say something like "if unable to determine the local or remote version, this does nothing"

File: src/checkFirebaseSDKVersion.ts [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+ * @param sourceDir Source directory of functions code
+ **/
+export function getFunctionsSDKVersion(sourceDir: string): string | void {
+  try {
+    const child = spawn.sync("npm", ["list", "firebase-functions", "--json=true"], {
+      cwd: sourceDir,
+      encoding: "utf8",
+    });
+    if (child.error) {
+      logger.debug(child.error.stack);
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:36]: Printing a message along with the stack can be useful

File: src/checkFirebaseSDKVersion.ts (hunk 21 above)
Comment (laurenzlong) [reply src/checkFirebaseSDKVersion.ts:71]: oh neat!

File: src/checkFirebaseSDKVersion.ts [hunk 24]
```diff
@@ -0,0 +1,104 @@
...
+interface NpmShowResult {
+  "dist-tags": {
+    latest: string;
+  };
+}
+
+/**
+ * Returns the version of firebase-functions SDK specified by package.json and package-lock.json.
+ * @param sourceDir Source directory of functions code
+ * @returns version string (e.g. "3.1.2"), or void if firebase-functions is not in package.json
```
Comment (bkendall) [inline src/checkFirebaseSDKVersion.ts:30]:
```suggestion
 * @return version string (e.g. "3.1.2"), or void if firebase-functions is not in package.json
```
